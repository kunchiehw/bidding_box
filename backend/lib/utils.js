const jwt = require('jsonwebtoken');

const secret = process.env.SHARE_SECRET;
const db = {
  users: {
    wkc: 'password',
    jarron: 'password',
  },
};


function authenticateUser(username, password) {
  return new Promise((resolve, reject) => {
    if (!username || !password || !(username in db.users) || db.users[username] !== password) {
      return reject();
    }

    const token = jwt.sign({ username }, secret, { expiresIn: '30d' });
    return resolve(token);
  });
}
module.exports.authenticateUser = authenticateUser;


function validateJwt(token) {
  return new Promise((resolve, reject) => {
    jwt.verify(token, secret, (err, decoded) => {
      if (err || !decoded) {
        reject();
      }

      if (!('username' in decoded) || !(decoded.username in db.users)) {
        reject();
      }

      resolve(token);
    });
  });
}
module.exports.validateJwt = validateJwt;


module.exports.validateJwtMiddleware = (req, res, next) => {
  if (req.headers.authorization && req.headers.authorization.split(' ')[0] === 'Bearer') {
    const token = req.headers.authorization.split(' ')[1];
    validateJwt(token)
      .then(() => next())
      .catch(() => next(new Error('Unauthorized')));
  } else {
    next(new Error('Unauthorized'));
  }
};


module.exports.getTtl = () =>
  Math.floor(Date.now() / 1000) + (4 * 60 * 60); // ttl for 4 hour
