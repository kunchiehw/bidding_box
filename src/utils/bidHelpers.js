const { STANDARD_SUITS } = require('./constants');

function isPass(bid) {
  return (bid && bid.suit === 'PASS');
}

function isDouble(bid) {
  return (bid && bid.suit === 'DOUBLE');
}

function isSuit(bid) {
  return (bid && STANDARD_SUITS.includes(bid.suit));
}

function isBidSeqEnded(bidSeq) {
  const bidSeqLen = bidSeq.length;
  if (isPass(bidSeq[bidSeqLen - 3]) &&
      isPass(bidSeq[bidSeqLen - 2]) &&
      isPass(bidSeq[bidSeqLen - 1])) {
    return (bidSeqLen >= 4);
  }
  return false;
}

function isDoubleAllowed(bidSeq) {
  const bidSeqLen = bidSeq.length;
  if (isSuit(bidSeq[bidSeqLen - 1])) {
    return false;
  }
  if (isSuit(bidSeq[bidSeqLen - 3]) &&
      isPass(bidSeq[bidSeqLen - 2]) &&
      isPass(bidSeq[bidSeqLen - 1])) {
    return false;
  }
  return true;
}

function isRedoubleAllowed(bidSeq) {
  const bidSeqLen = bidSeq.length;
  if (isDouble(bidSeq[bidSeqLen - 1])) {
    return false;
  }
  if (isDouble(bidSeq[bidSeqLen - 3]) &&
      isPass(bidSeq[bidSeqLen - 2]) &&
      isPass(bidSeq[bidSeqLen - 1])) {
    return false;
  }
  return true;
}

function getCurrentBid(bidSeq) {
  for (let i = bidSeq.length - 1; i >= 0; i -= 1) {
    if (isSuit(bidSeq[i])) {
      return bidSeq[i];
    }
  }
  return { level: 0, suit: 'PASS' };
}

function isStandardBidAllowed(currentLevel, currentSuit, level, suit) {
  if (currentLevel !== level) {
    return (currentLevel > level);
  }
  return (STANDARD_SUITS.indexOf(currentSuit) >= STANDARD_SUITS.indexOf(suit));
}


module.exports = {
  isPass,
  isDouble,
  isSuit,
  isBidSeqEnded,
  isDoubleAllowed,
  isRedoubleAllowed,
  getCurrentBid,
  isStandardBidAllowed,
};
