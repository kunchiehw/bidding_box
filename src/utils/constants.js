export const POSSIBLE_LEVELS = [1, 2, 3, 4, 5, 6, 7];

export const STANDARD_SUITS = ['CLUBS', 'DIAMONDS', 'HEARTS', 'SPADES', 'NOTRUMPS'];
export const SPECIAL_SUITS = ['PASS', 'DOUBLE', 'REDOUBLE'];
export const ALL_SUITS = ['CLUBS', 'DIAMONDS', 'HEARTS', 'SPADES', 'NOTRUMPS', 'PASS', 'DOUBLE', 'REDOUBLE'];

export const NS_SEATS = ['NORTH', 'SOUTH'];
export const EW_SEATS = ['WEST', 'EAST'];
export const SEATS = ['WEST', 'NORTH', 'EAST', 'SOUTH'];
export const DECLARERS = ['WEST', 'NORTH', 'EAST', 'SOUTH', 'NS', 'EW'];
export const PARTICIPANTS = ['WEST', 'EAST', 'WATCHER'];

export const VULNERABILITIES = ['NONE', 'NS', 'EW', 'BOTH'];

export const BOOLEANS = [true, false];

export const SERVER_API_HOST = `${process.env.REACT_APP_BACKEND_SCHEMA}://${process.env.REACT_APP_BACKEND_URL}`;
export const SERVER_WS_HOST = `ws://${process.env.REACT_APP_BACKEND_URL}`;
