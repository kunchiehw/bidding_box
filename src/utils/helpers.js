import { STANDARD_SUITS } from './constants';
import getMessage from '../i18n/messages';

export function bidToString(level, suit) {
  if (STANDARD_SUITS.includes(suit)) {
    return `${level} ${getMessage(suit)}`;
  }
  return getMessage(suit);
}

export function suitColor(suit) {
  switch (suit) {
    case 'NOTRUMPS':
      return 'black';
    case 'SPADES':
    case 'REDOUBLE':
      return 'blue';
    case 'HEARTS':
    case 'DOUBLE':
      return 'red';
    case 'DIAMONDS':
      return 'orange';
    case 'CLUBS':
    case 'PASS':
      return 'green';
    default:
      return '';
  }
}

export function shouldStandardBidButtonDisabled(currentLevel, currentSuit, buttonLevel, buttonSuit) {
  if (currentLevel !== buttonLevel) {
    return (currentLevel > buttonLevel);
  }
  return (STANDARD_SUITS.indexOf(currentSuit) >= STANDARD_SUITS.indexOf(buttonSuit));
}
