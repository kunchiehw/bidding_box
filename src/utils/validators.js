import { POSSIBLE_LEVELS, STANDARD_SUITS } from './constants';

// HOC for all validator
function createChainableTypeChecker(validate) {
  function checkType(isRequired, props, propName, componentName, location) {
    componentName = componentName || 'ANONYMOUS';
    if (props[propName] == null) {
      if (isRequired) {
        return new Error(`'${componentName}': requires prop '${propName}'.`);
      }
      return null;
    }
    return validate(props, propName, componentName, location);
  }

  const chainedCheckType = checkType.bind(null, false);
  chainedCheckType.isRequired = checkType.bind(null, true);

  return chainedCheckType;
}

export const levelValidator = createChainableTypeChecker((props, propName, componentName) => {
  componentName = componentName || 'ANONYMOUS';

  if (props.suit == null) {
    return new Error(`'${componentName}': requires prop 'suit' when validates prop '${propName}'.`);
  }

  if (props[propName]) {
    const level = props[propName];
    if (STANDARD_SUITS.includes(props.suit)) {
      if (!POSSIBLE_LEVELS.includes(level)) {
        return new Error(`'${componentName}': '${propName}' should be in POSSIBLE_LEVELS when 'suit' is in STANDARD_SUITS.`);
      }
    } else if (level !== 0) {
      return new Error(`'${componentName}': '${propName}' should be 0 when 'suit' is in SPECIAL_SUITS.`);
    }
  }

  return null;
});

export function foo() {}
