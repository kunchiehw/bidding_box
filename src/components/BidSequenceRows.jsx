import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Segment } from 'semantic-ui-react';
import { SEATS, STANDARD_SUITS, ALL_SUITS } from '..//utils/constants';
import { bidToString, suitColor } from '../utils/helpers';
import { levelValidator } from '../utils/validators';

const propTypes = {
  dealer: PropTypes.oneOf(SEATS).isRequired,
  bidSeq: PropTypes.arrayOf(PropTypes.shape({
    level: levelValidator.isRequired,
    suit: PropTypes.oneOf(ALL_SUITS).isRequired,
  })).isRequired,
};

function BidSequenceRows(props) {
  const bidSequenceRows = [];
  let bidSequenceRow = [];
  let bidSequenceRowIndex = 0;

  // Generate empty cells first when the dealer is not WEST (the left most player).
  for (let seatIndex = 0; seatIndex < SEATS.length; seatIndex += 1) {
    if (props.dealer !== SEATS[seatIndex]) {
      bidSequenceRow.push((
        <Grid.Column key={SEATS[seatIndex]}>
          <Segment className="bid-sequence-cell empty-cell">
            Empty
          </Segment>
        </Grid.Column>
      ));
    } else {
      break;
    }
  }

  for (let bidSeqIndex = 0; bidSeqIndex < props.bidSeq.length; bidSeqIndex += 1) {
    const bid = props.bidSeq[bidSeqIndex];
    if (STANDARD_SUITS.includes(bid.suit)) {
      bidSequenceRow.push((
        <Grid.Column key={bidSeqIndex}>
          <Segment className="bid-sequence-cell" color={suitColor(bid.suit)} inverted>
            {bidToString(bid.level, bid.suit)}
          </Segment>
        </Grid.Column>
      ));
    } else {
      bidSequenceRow.push((
        <Grid.Column key={bidSeqIndex} >
          <Segment className={`bid-sequence-cell ${bid.suit}-cell`}>
            {bidToString(bid.level, bid.suit)}
          </Segment>
        </Grid.Column>
      ));
    }
    if (bidSequenceRow.length === 4) {
      bidSequenceRows.push((
        <Grid.Row key={bidSequenceRowIndex} className="bid-sequence-row" column={4}>
          {bidSequenceRow}
        </Grid.Row>
      ));
      bidSequenceRow = [];
      bidSequenceRowIndex += 1;
    }
  }

  bidSequenceRows.push((
    <Grid.Row key={bidSequenceRowIndex} className="bid-sequence-row" column={4}>
      {bidSequenceRow}
    </Grid.Row>
  ));

  return bidSequenceRows;
}

BidSequenceRows.propTypes = propTypes;

export default BidSequenceRows;
