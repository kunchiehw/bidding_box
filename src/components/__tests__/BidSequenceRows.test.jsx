import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import BidSequenceRows from '../BidSequenceRows';
import { SEATS } from '../../utils/constants';

describe('BidSequenceRows render', () => {
  const bidSeqs = [
    [],
    [
      { level: 1, suit: 'CLUBS' },
      { level: 2, suit: 'DIAMONDS' },
      { level: 3, suit: 'HEARTS' },
      { level: 4, suit: 'SPADES' },
      { level: 5, suit: 'NOTRUMPS' },
    ],
    [
      { level: 6, suit: 'CLUBS' },
      { level: 0, suit: 'DOUBLE' },
      { level: 0, suit: 'REDOUBLE' },
      { level: 0, suit: 'PASS' },
      { level: 0, suit: 'PASS' },
      { level: 0, suit: 'PASS' },
    ],
  ];

  SEATS.forEach((seat) => {
    bidSeqs.forEach((bidSeq) => {
      it(`seat: ${seat}, bidSeq: ${bidSeq}`, () => {
        const bidSequenceRows = shallow(<BidSequenceRows
          dealer={seat}
          bidSeq={bidSeq}
        />);
        expect(toJson(bidSequenceRows)).toMatchSnapshot();
      });
    });
  });
});
