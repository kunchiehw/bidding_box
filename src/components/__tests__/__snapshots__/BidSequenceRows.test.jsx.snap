// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BidSequenceRows render seat: EAST, bidSeq:  1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
]
`;

exports[`BidSequenceRows render seat: EAST, bidSeq: [object Object],[object Object],[object Object],[object Object],[object Object] 1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="1"
/>,
]
`;

exports[`BidSequenceRows render seat: EAST, bidSeq: [object Object],[object Object],[object Object],[object Object],[object Object],[object Object] 1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="1"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="2"
/>,
]
`;

exports[`BidSequenceRows render seat: NORTH, bidSeq:  1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
]
`;

exports[`BidSequenceRows render seat: NORTH, bidSeq: [object Object],[object Object],[object Object],[object Object],[object Object] 1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="1"
/>,
]
`;

exports[`BidSequenceRows render seat: NORTH, bidSeq: [object Object],[object Object],[object Object],[object Object],[object Object],[object Object] 1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="1"
/>,
]
`;

exports[`BidSequenceRows render seat: SOUTH, bidSeq:  1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
]
`;

exports[`BidSequenceRows render seat: SOUTH, bidSeq: [object Object],[object Object],[object Object],[object Object],[object Object] 1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="1"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="2"
/>,
]
`;

exports[`BidSequenceRows render seat: SOUTH, bidSeq: [object Object],[object Object],[object Object],[object Object],[object Object],[object Object] 1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="1"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="2"
/>,
]
`;

exports[`BidSequenceRows render seat: WEST, bidSeq:  1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
]
`;

exports[`BidSequenceRows render seat: WEST, bidSeq: [object Object],[object Object],[object Object],[object Object],[object Object] 1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="1"
/>,
]
`;

exports[`BidSequenceRows render seat: WEST, bidSeq: [object Object],[object Object],[object Object],[object Object],[object Object],[object Object] 1`] = `
Array [
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="0"
/>,
  <GridRow
    className="bid-sequence-row"
    column={4}
    key="1"
/>,
]
`;
